_posts/2023-07-07-python-jails.md

```diff
@@ -0,0 +1,245 @@
+---
+layout: post
+title: "Python Jail Escapes"
+categories: learning
+author: Liam
+---
+
+At our weekly meetings we had a talk about Python jail escapes, aka. getting around restrictions that make it hard to execute `os.system('cat flag.txt')`. In the talk we went through challenges, that we present here as exercises to practice. Starting very simple and then digging more and more into python internal.
+
+Solve them yourself, by copying the script and providing input that executes `os.system('cat flag.txt')`. Afterwards you can compare your solution to the one we had in mind, there are always different ways. For some reference solution it is relevant to mention, that we used Python 3.11.2.
+
+Leaning about Python Jail Escapes it is helpful to ask questions like:
+* How are modules and classes implemented in Python?
+* What are frozen modules?
+* How does syntactic sugar get replaced? You may start at `import someting` or `array[5]`.
+* How many ways are there to call a function/method other than `lmao()`/`lm.oa()`?
+
+## Happens way too often
+
+```python
+import string
+
+code = input('Your scientific computation: ')
+
+code = ''.join([c for c in code if c in string.printable])
+
+forbidden = ['eval', 'exec', 'import', 'open', 'system', 'os', 'builtins']
+for f in forbidden:
+    code = code.replace(f, '')
+exec(code)
+```
+
+<details>
+<summary>Reference Solution</summary>
+
+No need for recursivity. This is Python not Haskell.
```

Comment:
recursiveness

Reply:
It's fine as is, if we want to believe Google Ngram:
https://books.google.com/ngrams/graph?content=recursivity%2Crecursiveness&year_start=1800&year_end=2019&corpus=en-2019&smoothing=3

Reply:
Really?! Wow...